lxd/db/cluster/entities.go

```diff
@@ -81,3 +85,49 @@ func init() {
 		EntityTypes[name] = code
 	}
 }
+
+// URLToEntityType parses a raw URL string and returns the entity type, the project (if present), and the
+// path arguments. The returned project is not automatically set to "default", this is the responsibility of the caller.
+// An error is returned if the URL is not recognised.
+func URLToEntityType(rawURL string) (int, string, []string, error) {
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return -1, "", nil, err
+	}
+
+	// We need to space separate the path because fmt.Sscanf uses this as a delimiter.
+	spaceSeparatedURLPath := strings.Replace(u.Path, "/", " / ", -1)
+	for entityType, entityURI := range EntityURIs {
+		entityPath, _, _ := strings.Cut(entityURI, "?")
+
+		// Skip if we don't have the same number of slashes.
+		if strings.Count(entityPath, "/") != strings.Count(u.Path, "/") {
+			continue
+		}
+
+		spaceSeparatedEntityPath := strings.Replace(entityPath, "/", " / ", -1)
+
+		// Make an []any for the number of expected path arguments and set each value in the slice to a *string.
+		nPathArgs := strings.Count(spaceSeparatedEntityPath, "%s")
+		pathArgsAny := make([]any, nPathArgs)
+		for i := range pathArgsAny {
+			var pathComponentStr string
+			pathArgsAny[i] = &pathComponentStr
+		}
+
+		// Scan the given URL into the entity URL. If we found all the expected path arguments and there
+		// are no errors we have a match.
+		nFound, err := fmt.Sscanf(spaceSeparatedURLPath, spaceSeparatedEntityPath, pathArgsAny...)
+		if nFound == nPathArgs && err == nil {
+			pathArgs := make([]string, nPathArgs)
+			for i, pathArgAny := range pathArgsAny {
+				pathArgPtr := pathArgAny.(*string)
+				pathArgs[i] = *pathArgPtr
+			}
+
+			return entityType, u.Query().Get("project"), pathArgs, nil
```

Comment:
Should this return "default" for project if not present in URL query string?

Reply:
I've set the project to default if it is not present. It should be unused if the returned entity type isn't project specific. Additionally I've set the project to the project path parameter if the entity type is `TypeProject` in order to avoid confusion.
I've added a comment to explain the above also.